ray-operator/test/e2eautoscaler/raycluster_autoscaler_part2_test.go

```diff
@@ -500,3 +501,101 @@ func TestRayClusterAutoscalerPlacementGroup(t *testing.T) {
+
+			rayClusterSpecAC := rayv1ac.RayClusterSpec().
+				WithEnableInTreeAutoscaling(true).
+				WithGcsFaultToleranceOptions(
+					rayv1ac.GcsFaultToleranceOptions().
+						WithRedisAddress(RedisAddress).
+						WithRedisPassword(rayv1ac.RedisCredential().WithValue(RedisPassword)),
+				).
+				WithRayVersion(GetRayVersion()).
+				WithHeadGroupSpec(rayv1ac.HeadGroupSpec().
+					WithRayStartParams(map[string]string{
+						"num-cpus": "0",
+					}).
+					WithTemplate(tc.HeadPodTemplateGetter()),
+				).
+				WithWorkerGroupSpecs(rayv1ac.WorkerGroupSpec().
+					WithRayStartParams(map[string]string{
+						"num-cpus": "1",
+					}).
+					WithGroupName("small-group").
+					WithReplicas(0).
+					WithMinReplicas(0).
+					WithMaxReplicas(2).
+					WithTemplate(tc.WorkerPodTemplateGetter()),
+				)
+			rayClusterAC := rayv1ac.RayCluster("ray-cluster", namespace.Name).
+				WithSpec(apply(rayClusterSpecAC, mountConfigMap[rayv1ac.RayClusterSpecApplyConfiguration](scripts, "/home/ray/test_scripts")))
+
+			rayCluster, err := test.Client().Ray().RayV1().RayClusters(namespace.Name).Apply(test.Ctx(), rayClusterAC, TestApplyOptions)
+			g.Expect(err).NotTo(gomega.HaveOccurred())
+			LogWithTimestamp(test.T(), "Created RayCluster %s/%s successfully", rayCluster.Namespace, rayCluster.Name)
+
+			// Wait for RayCluster to become ready and verify the number of available worker replicas.
+			g.Eventually(RayCluster(test, rayCluster.Namespace, rayCluster.Name), TestTimeoutMedium).
+				Should(gomega.WithTransform(RayClusterState, gomega.Equal(rayv1.Ready)))
+			g.Expect(GetRayCluster(test, rayCluster.Namespace, rayCluster.Name)).To(gomega.WithTransform(RayClusterDesiredWorkerReplicas, gomega.Equal(int32(0))))
+
+			headPod, err := GetHeadPod(test, rayCluster)
+			g.Expect(err).NotTo(gomega.HaveOccurred())
+			LogWithTimestamp(test.T(), "Found head pod %s/%s", headPod.Namespace, headPod.Name)
+
+			// Create a detached actor, and a worker should be created.
+			ExecPodCmd(test, headPod, common.RayHeadContainer, []string{"python", "/home/ray/test_scripts/create_detached_actor.py", "actor1"})
+			g.Eventually(RayCluster(test, rayCluster.Namespace, rayCluster.Name), TestTimeoutMedium).
+				Should(gomega.WithTransform(RayClusterDesiredWorkerReplicas, gomega.Equal(int32(1))))
+
+			// Terminate a detached actor, and a worker should be deleted.
+			ExecPodCmd(test, headPod, common.RayHeadContainer, []string{"python", "/home/ray/test_scripts/terminate_detached_actor.py", "actor1"})
+			g.Eventually(RayCluster(test, rayCluster.Namespace, rayCluster.Name), TestTimeoutMedium).
+				Should(gomega.WithTransform(RayClusterDesiredWorkerReplicas, gomega.Equal(int32(0))))
+
+			// Delete the head Pod
+			err = test.Client().Core().CoreV1().Pods(namespace.Name).Delete(test.Ctx(), headPod.Name, metav1.DeleteOptions{})
+			g.Expect(err).NotTo(HaveOccurred())
+
+			PodUID := func(p *corev1.Pod) string { return string(p.UID) }
```

Comment:
maybe we can add a util function for 

```go
			PodUID := func(p *corev1.Pod) string { return string(p.UID) }
			g.Eventually(HeadPod(test, rayCluster), TestTimeoutMedium).
				ShouldNot(WithTransform(PodUID, Equal(string(headPod.UID)))) // Use UID to check if the new head pod is created.

			g.Eventually(HeadPod(test, rayCluster), TestTimeoutMedium).
				Should(WithTransform(func(p *corev1.Pod) string { return string(p.Status.Phase) }, Equal("Running")))

			headPod, err = GetHeadPod(test, rayCluster) // Replace the old head pod
			g.Expect(err).NotTo(HaveOccurred())
```

I guess the logic is used more than once in the codebase?

Reply:
You can either update it in this PR or open an issue and ask other contributors for help.

Reply:
Add a task here https://github.com/ray-project/kuberay/issues/3889